### PR TITLE
Version 2 roadmap

### PR DESCRIPTION
**Pull request by sukima**
This is an all encompassing PR to cover #10 and #11 as well as address issues #7, #8, #12, and #13

It is designed to be merged for a **Breaking Change** version bump (`2.0.0`).

It introduces `validateDidInsert`, `validateTracked`, capturing all errors, and handling event bubbling (attach to non-field elements).

- [x] Write docs for `validateDidInsert`
- [x] Write docs for `validateTracked`
- [x] Write docs for `validated` event detail (multiple errors handling)
- [x] Add doc statement about bubbling for non-field elements
- [x] Write docs for `verify-form-validity`
- [x] Write tests for `verify-form-validity`
- [x] Build out a more user friendly demo/documentation site in dummy app and deploy to gh-pages
- [x] Expand `validateTracked` semantics and reasons for it
- [x] Clean Git history

### CODE REVIEW

File: addon/-private/auto-tracking-exerciser.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  }
+
+  runImmediatlyWhen(runImmediatly) {
+    this.runImmediatly = runImmediatly;
+    return this;
+  }
+
+  exercise(callback) {
+    this.props.forEach(i => i);
+    console.count('exercise');
```
Comment (sukima) [inline addon/-private/auto-tracking-exerciser.js]: Remove this

File: tests/dummy/app/controllers/index-example.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  @action
+  async handleFormSubmit(event) {
+    event.preventDefault();
+    let { target: form } = event;
+    await validate(...form.elements);
+    if (form .checkValidity()) {
+      this.createUser(Object.fromEntries(new FormData(form)));
+      this.resetForm(form);
+    }
+  }
```
Comment (sukima) [inline tests/dummy/app/controllers/index-example.js:23]: Should we add a modifier to do this default on form elements?

```hbs
<form {{verify-validity submit=this.submit}}>
  …
</form>
```

File: tests/dummy/app/controllers/index-example.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      [target.name]: target.validationMessage
+    };
+  }
+
+  @action
+  async handleFormSubmit(event) {
+    event.preventDefault();
+    let { target: form } = event;
+    await validate(...form.elements);
+    if (form .checkValidity()) {
```
Comment (sukima) [inline tests/dummy/app/controllers/index-example.js]: Extra space typo

File: tests/dummy/app/controllers/index-example.js (hunk 2 above)
Comment (sukima) [reply tests/dummy/app/controllers/index-example.js:23]: Added

**Comment by sukima** [comment]
I just tried this branch in one of our documentation projects and it works great! I think I'm going to clean this up and merge soon.

**Comment by sukima** [comment]
Spoke too soon. using `validateDidInsert` caused an error

> Uncaught (in promise) TypeError: target.setCustomValidity is not a function

**Comment by sukima** [comment]
Because I set this up to work with event bubbling it enabled the modifier to be placed on non-field elements. This meant that when the user triggered a validation (via input/change/blur) it would bubble and trigger the handler. Since the original element was the input element the `event.target` is a valid input element and thus has a `setCustomValidity()`.

When I performed this with the `validateDidInsert` flag the modifier has no idea of the element it is attached to thus dispatching the validate event on a non-input element. When the handler triggers the `event.target` is not an input and thus `setCustomValidity()` is not available.

I need to figure out either placing an assert to prevent this (or offer a better error message) OR maybe instead attempt to discover the input element via a DOM walk?

**Comment by sukima** [comment]
Fixed missing`setCustomValidity` error with `validateDidInsert`/`validateTracked` use. Easy enough to walk the DOM to find valid input fields. And if none found no-op.

**Comment by timlindvall** [comment]
Took a look at the diff and the changes LGTM! I'm gonna test this out locally a little more - I'll post an update if I find anything from that.

Update: No issues testing locally either. I'll take another pass at the diff, but don't think I'll come across anything glaring. Thanks!

File: README.md [hunk 4]
```diff
@@ -71,14 +71,34 @@ By default validation will happen on `change`, `input`, and `blur`. Comma separa
...
+</div>
+```
+
 #### Example validation on form submit
 
 ```hbs
-<form ...attributes novalidate {{on "submit" this.validateAndSubmit}}>
+<form
+  ...attributes
+  {{verify-form-validity submit=this.validateAndSubmit reportValidity=true}}
```
Comment (timlindvall) [inline README.md]: nit: Consider renaming `this.validateAndSubmit` now that the function no longer triggers validation.

File: tests/dummy/app/templates/index-example.hbs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    >
+    <span class="validation-message">{{this.validationMessages.password}}</span>
+
+    <label for="userForm-confirm">Confirm password<sup class="required">*</sup></label>
+    <input
+      id="userForm-confirm"
+      name="confirm"
+      type="password"
+      required
+      {{validity (validate-confirmation "userForm-password")}}
```
Comment (timlindvall) [inline tests/dummy/app/templates/index-example.hbs]: For demo purposes, assuming the password field set some property on the backing controller on input, this could leverage `validateTracked` to error when the first field changes.

File: tests/dummy/app/templates/index-example.hbs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    <span class="validation-message">{{this.validationMessages.name}}</span>
+
+    <label for="userForm-phone">Phone</label>
+    <input id="userForm-phone" name="phone" type="tel" {{validity}}>
+    <span class="validation-message">{{this.validationMessages.phone}}</span>
+
+    <label for="userForm-email">Email<sup class="required">*</sup></label>
+    <input id="userForm-email" name="email" type="email" required {{validity}}>
+    <span class="validation-message">{{this.validationMessages.email}}</span>
+
```
Comment (timlindvall) [inline tests/dummy/app/templates/index-example.hbs]: Add a select for demo purposes? (idk, something like "Preferred contact method" to stick with the demo theme?)

**Review by timlindvall — COMMENTED** [review]
minor nitpicks to consider if it helps. again, though, looks good overall!

File: README.md [hunk 7]
```diff
@@ -89,35 +109,49 @@ import { validate } from 'ember-validity-modifier';
...
+    console.log('Fake submit action', Object.fromEntries(new FormData(form)));
   }
 }
 ```
 
-#### Example with `validator-update` event
+#### Example with `validateDidInsert` argument
 
-To validate the form state on initial render and any time its dependent arguments change, add the `'validator-update'` event to the list of events passed in via the `on` argument.
+To validate the form state on initial render add `validateDidInsert=true`.
```
Comment (NullVoxPopuli) [inline README.md]: Is there another name that'd be better suited here?
I think it'd be great if render lifecycle was totally abstracted away from the end user

File: README.md [hunk 8]
```diff
@@ -89,35 +109,49 @@ import { validate } from 'ember-validity-modifier';
...
 
 ```hbs
-  <input {{validity (fn this.matchTo this.match) on="change,validator-update"}}>
+  <input {{validity
+    (fn this.matchTo this.match)
+    on="change"
+    validateDidInsert=true
```
Comment (NullVoxPopuli) [inline README.md]: Maybe eager?

File: README.md [hunk 9]
```diff
@@ -89,35 +109,49 @@ import { validate } from 'ember-validity-modifier';
...
+
+Because the validator argument is a function it is possible to not exercise the tracked properties and thus miss out on validations when those tracked properties change. This is the case of the `fn` helper which lazy executes thus doesn't trigger Ember's auto-tracking if it isn't ran first.
+
+To compensate we can use `validateTracked` to inform the modifier that it needs to run the validations when these properties change.
+
+```hbs
+  <input {{validity
+    (fn this.matchTo this.match)
+    on="change"
+    validateTracked=this.match
```
Comment (NullVoxPopuli) [inline README.md:145]: This shouldn't be needed. As long as the modifier consumes all args when it's installed, they'll be entangled

File: tests/dummy/app/templates/index-example.hbs (hunk 5 above)
Comment (sukima) [reply tests/dummy/app/templates/index-example.hbs]: That would only be a need if something outside this component modified the tracked property. Also this example has no backing data as it relies on the native FormData when the form submits.

Are you suggesting we use a different example to demonstrate the weirdness of Data Up/Down dynamics of tracked properties?

File: README.md (hunk 7 above)
Comment (sukima) [reply README.md]: How does the end user know this runs the validation immediately on the modifier function completing? Most users I know think of that as a life cycle. When this element is rendered (A.K.A. did insert) do this thing.

Do you have another suggestion?

File: README.md (hunk 8 above)
Comment (sukima) [reply README.md]: We tried eager in the past. Every time it was used I noticed myself and others confused by it. I think the term is to broad to be useful in this context.

File: README.md (hunk 9 above)
Comment (sukima) [reply README.md:145]: There is no way for the modifier to know what to consume as validations are lazy evaluated. So are helpers that return a function. The only way to have the modifier consume the properties is to list them explicitly which is what validatedTracked is all about.

Comment (sukima) [reply README.md:145]: The only other option would be to force all validation functions to execute and ignore the results. But that would add a level of confusion to consumers to see their validator functions execute for no reason.

File: README.md (hunk 7 above)
Comment (sukima) [reply README.md]: `validateFirstRender`?

Comment (NullVoxPopuli) [reply README.md]: maybe sub-object?
```
validate=(hash
  eagerly=true
)
```

File: README.md (hunk 9 above)
Comment (NullVoxPopuli) [reply README.md:145]: oh interesting, I didn't know `fn` disconnects the reactivity :thinking: 

I'd probably switch to this api then:
https://github.com/NullVoxPopuli/tracked-modifier-demo/commit/1efc6f626f974f882a74169e7044e64767942863#diff-3537d24c7c7ac3a6699672257220a629b4f13066903e5b8b206097717f9d12e0R2

Comment (sukima) [reply README.md:145]: `fn` seems to be lazy and will only register auto-tracking if it is executed.

Can you explain the demo api. I don't understand it. Also I expect this addon should remain backwards compatible to some older versions of Ember.

Comment (NullVoxPopuli) [reply README.md:145]: that api is 3.25+, so it should be fine.

The gist is by _not_ using `fn`, the positional args' "trackedness" are consumed, so when `match` changes, the modifier tears itself down, and sets itself back up again -- eliminating the need in this PR to have `validateTracked`

Comment (sukima) [reply README.md:145]: If I understand you are suggesting changing the `validity` modifier API to accept only *one* validator function and the rest of the params are curried args to the validator function. Thus triggering the auto-tracking?

I see where that goes but I know I and others have asked for the API that each params is it's own validator function. I presume others being familiar with the `(fn)` helper in that case. Or in that use of multiple helpers.

```hbs
<input {{validity
  (validate-something)
  (validate-something-else this.foobar)
}}>
```

This could be done by checking the `typeof` of each param. However it still doesn't help the use case where a user wants to pass in a function via a helper and also have the modifier validate when something on that helper changes. Both helpers and `(fn)` are lazy like that. For it to work everyone who writes a validator helper must also exercise the params for auto-tracking. This is a nuance that I think would be an undo burden to end users.

In other words example 1 is easier for end users to make their own helpers then example 2:

#### Example 1

```js
export default Helper.helper(function (params) {
  return () => handleParamsButLazyInsideClosureFunction(params) ? [] : ['validation message'];
});
```

#### Example 2

```js
export default Helper.helper(function (params) {
  params.forEach(i => i); // All users have to remember to do this or things break.
  return () => handleParamsButLazyInsideClosureFunction(params) ? [] : ['validation message'];
});
```

Comment (NullVoxPopuli) [reply README.md:145]: helpers should also auto-track like if those validators follow the function-factory pattern.

I imagine you could provide a wrapper function that each validator could use, like:

```js
export const validateSomething = validator((/* no args */) => {
  
});

export const validateSomethingElse = validator(foobar => {

});
```
where _validator_ does the entaglement from the demo

Comment (sukima) [reply README.md:145]: @timlindvall What do you think about the idea that the act of adding the modifier will execute each validator function even if the result is ignored? If that seems like a reasonable trade off then I think I can accomplish @NullVoxPopuli proposal to drop the `validateTracked` API.

I take that back. validators could be asynchronous and possibly make network requests. That would be bad to execute them without any need to.

Comment (NullVoxPopuli) [reply README.md:145]: >  will execute each validator function

I want to be _extra_ clear that this assumes that the validator functions would have to become factories that invoke the passed callback to `validator` when it's time to validate

File: README.md (hunk 7 above)
Comment (sukima) [reply README.md]: Almost the same as the `on=` parameter.

`validateImmediatly=true`?

File: README.md (hunk 9 above)
Comment (sukima) [reply README.md:145]: That's doable. I wonder if there is a way to tell the difference between a factory and a normal function. If so then we could have best of both worlds. 🤔

Comment (NullVoxPopuli) [reply README.md:145]: you could use trickyJS and say `validator.__factory = true` and check for the presence of that property?

Comment (timlindvall) [reply README.md:145]: Sorry, catching up on this thread. 😅 

It should be ok to have `{{validate}}` only accept one validation function per usage if we want to go that route. If someone wants multiple validators, they can attach multiple `{{validate}}` modifiers, which should collapse into one validation check. (And, if someone wants to pass in another validation method to a sub-component, composing using `...attributes` is a suitable option.)

Or, alternatively, the helper-function-factory pattern seemed to work well when I was working with the modifier, but I was using `validator-update` so it's possible that's because I was validating immediately. Though it would make sense to me that any props passed to a helper _should_ trigger autotracking. But... it carries the same caveat that you got at above... if I want to track something, I have to pass it in as an arg, even if I don't care about its value. (Perhaps that suggests a refactoring opportunity for that particular piece of code, though.)

Comment (NullVoxPopuli) [reply README.md:145]: oh multiple `{{validate}}` should probably work :thinking: good thinking!

Comment (sukima) [reply README.md:145]: I think there are two issues. Multiple modifiers on one element is already supported by this branch. Thus the restriction to only support one and let suffix params be args to the validator (partial application) is possible.

However, this won't help in the case of a user wishing to use the `(fn)` helper. the only way the `(fn)` helper will auto track is if the function is executed. Normally in the case of callbacks (actions up) this isn't really an issue. But it does become an issue if the function is not executed but deferred to another execution later like we do for validations since validators can be asynchronous.

We have three options I see

1. support `(fn)` via a `validateTracked` property —Needs documentation and might look odd
2. reinvent `(fn)` internally such that `[validator, ...argsForValidatorPartialApplication]` is the modifier's interface. —Users would be unaware of the caveats if they use `(fn)`
3. Provide mechanism for defining validator factories —adds extra complexity and still has the issue that users need to be aware of auto-tracking from a programming stand point.

I hate to say it but I'm still in the `validateTracked` camp. mainly because it is the least resistance to using the `(fn)` helper (or custom helpers without needing to exercise the auto-tracking that users will inevitably forget.

My understand so far is that the issue is that we are fighting two paradigms. Normally when something like a curried function is passed into something like a modifier it doesn't matter what it is. Only at execution time does it matter. In fact in the case of this modifier it also doesn't matter.

But it does matter if the tracked properties used to construct that function change and something unrelated (the act of validation) is dependent on that. Thus when you look at this situation both the validator *and* the modifier itself are dependents of the tracked properties.

By the design of tracked properties the thing that depends on them needs to use the. This is true for the `(fn)` helper and in our case true for the modifier itself.

That is the driving need for the `validateTracked` as it does the same thing as passing them as params though it is more explicit in this case. Here are two examples on how a modifier is intended to interact with tracked properties in terms of this addon:

#### Explicit

```hbs
<input {{validity (fn this.foo this.bar) validateTracked=(array this.foo this.bar)}}>
```

This version allows multiple validators in one modifier.

```hbs
<input {{validity
  (validate-foo this.foo)
  (validate-bar this.bar)
  validateTracked=(array this.foo this.bar)
}}>
```

#### Implicit

```hbs
<input {{validity (fn this.foo this.bar) this.foo this.bar}}>
```

This version forces only one validator per modifier.

```hbs
<input
  {{validity (validate-foo this.foo) this.foo}}
  {{validity (validate-bar this.bar) this.bar}}
>
```

Question is which one is easier better for users?

File: tests/dummy/app/templates/index-example.hbs (hunk 5 above)
Comment (timlindvall) [reply tests/dummy/app/templates/index-example.hbs]: _this might be moot based on the below thread, but, just in case..._

Yeah, this would suggest inventing a tracked property (or two) that the password field syncs to on input. As it is now, if the user were to set both password fields to the same, then change the first password field (thereby causing both to no longer equal), the second doesn't error until the next input/change/submit event.

It's a bit of a contrived example, but it demonstrates how to handle keeping two things (which could be two components coordinating through DDAU) in lockstep with each other or responding to changes somewhere else.

File: README.md (hunk 9 above)
Comment (sukima) [reply README.md:145]: What about a `tracked-validator` helper as a replacement for `fn` that actually exercises the passed args?

I'm experimenting with ideas. Or should we leave validateTracked as an optional API for those who like that syntax better?

This is a tough one as there really is only like three users of this addon. I don't know why good ideas don't spread.

Comment (NullVoxPopuli) [reply README.md:145]: Imo, the best api is no api. Increased surface area is room for confusion.

Comment (NullVoxPopuli) [reply README.md:145]: > Question is which one is easier better for users?

In your second example, you shouldn't need to duplicate the args

Comment (NullVoxPopuli) [reply README.md:145]:
```hbs
<input
  {{validity (validate-foo this.foo)}}
  {{validity (validate-bar this.bar)}}
>
```
Tho, i'd argue for this, since () tends to mean invocation:
```hbs
<input
  {{validity validate-foo this.foo}}
  {{validity validate-bar this.bar}}
>
```

Comment (sukima) [reply README.md:145]: Did you mean—

```hbs
<input
  {{validity (validate-foo) this.foo}}
  {{validity (validate-bar) this.bar}}
>
```

----

I think you are missing the point. You have to duplicate the args unless you can force all users who write custom helpers to exercise the args themselves. I think that ask is harder. And that users will easily be confused why the validation doesn't happen when their helper's args change. Also know that the act of validating is not by default when args change that would make things happen differently then normal forms do. It needs to be an opt-in situation.

Comment (sukima) [reply README.md:145]: The only way to do this with deterministic results is to execute the validator functions in the modifier so that the auto-tracking gets exercised But that means validator functions can no longer be asynchronous which is a want by many who need validation. See ember-changeset-validations which use async validators.

In the case of async validators someone has to have a list of args to exercise. Either we push that responsibility to the users who will get confused (if this.get/set was confusing imagine `(i) => i` in every helper). Or we take on the responsibility ourselves and provide an opt-in API surface to accommodate.

I believe this is the exact issue @timlindvall ran into and why they were motivated to provide a `validate-update` and eventually `validateTracked` API.

Comment (sukima) [reply README.md:145]: Thank you @NullVoxPopuli and @timlindvall your insights have been invaluable. I think for now I'm going to take on the `validateTracked` API to remain unblocked and we can iteratively introduce a better (less surface area) implementation next version.

Unless, @timlindvall you feel that this syntax is just as understandable I can do that instead:

```hbs
<input
  {{! One validity modifier per validator}}
  {{validity (validity-helper) this.arg1 this.arg2}}
  {{validity this.validatorFn this.arg1 this.arg2}}
>
```

The caveat that if someone uses `(fn this.args1 this.args2)` then updating arg1 and arg2 will **not** cause validations to be checked. Oof now that I wrote that I feel like any attempt to be "auto-magical" about tracking will only confuse the situation further. Least with `validateTracked` it may be over verbose but it carries with it both an opt-in semantic and is explicit to when it will perform a validation when args change.

Comment (sukima) [reply README.md:145]: I also think it might be worth while to draft a RFC like proposal to expose some of the tracked primitives so that addons can tap into the internals maybe offer some kind of escape hatch for situations like this.

File: README.md [hunk 10]
```diff
@@ -89,35 +109,49 @@ import { validate } from 'ember-validity-modifier';
...
-```js
-export default MyComponent extends Component {
-  @action
-  matchTo(match, { value }) {
-    return value === match ? [] : ['Must match exactly'];
-  }
-}
+#### Example with `validateTracked` argument
+
+To validate the form state when  initial render and any time **one** of its dependent arguments change, add the `'watch'` argument with the dependent property.
```
Comment (sukima) [inline README.md]: TODO: Expand more on why this is here and when to use it.

File: README.md (hunk 9 above)
Comment (NullVoxPopuli) [reply README.md:145]: > I think you are missing the point. You have to duplicate the args unless you can force all users who write custom helpers to exercise the args themselves. I

nay, I think I can maybe whip up a demo -- since worknig code is more effective than words.

File: tests/dummy/app/templates/index-example.hbs (hunk 5 above)
Comment (sukima) [reply tests/dummy/app/templates/index-example.hbs]: Ahh I see what you mean. Yes that could be managed via tracked properties and the `validateTracked` but it could also be managed vanilla by making the password also have a validation to check the reverse case. Updating example.

Comment (sukima) [reply tests/dummy/app/templates/index-example.hbs]: I played with this myself and I think you are right. The password/confirmation is a prime example of needing a tracked property and to validate on it being dirtied. I hope my helper `chage-tracker` illustrates this unique but necessary situation.

The only alternative I could imagine is to provide a `validate` modifier which could dispatch the `validate` DOM event but at that point I think the addon expands too far from the Ember expectations and ecosystem. I think that the business logic of related validators is best kept within the system it is being used. VanillaJS can use the `validate` function while Ember can use tracked properties through the `validateTracked` param.

I think I'm ok with the API as it is currently in this PR. We can refine this later.

I also think this situation and a few other threads here illustrate the power of the underlying design of DOM events and custom validity that it is flexible enough to have more then one way to skin the cat. 😉
